### PR TITLE
Added DuckDuckGo functionality

### PR DESCRIPTION
**Pull request by ManagarmrWhisperer**
### Yet to implement:

- [ ] Test all edge cases.
- [ ] Find selectors automatically if the ones we have fail, E.G Different themes.
- [ ] Fix badge misalignment.
- [ ] Properly test filtering.

### CODE REVIEW

File: js/ddg.js [hunk 1]
```diff
@@ -14,19 +15,223 @@ const wikis = prepareWikisInfo( getWikis( false, true ), {
...
+
+	const isTopLevel = (a) => {
+	    return a.href.startsWith(
+		`https://${wiki.oldId || wiki.id}.fandom.com`);
+	};
+	
+	// Check all a relements in the result and verify that it's a link to the same wiki
+	
+	for ( const a of element.getElementsByTagName('a')) {
+	    if(isTopLevel(a))
```
Comment (alex4401) [inline js/ddg.js:183]: Add spaces after `if`, both `(`, and before both `)`. Use braces. I'm following https://www.mediawiki.org/wiki/Manual:Coding_conventions/JavaScript as I'm used to do from MW extensions.

File: js/ddg.js [hunk 2]
```diff
@@ -14,19 +15,223 @@ const wikis = prepareWikisInfo( getWikis( false, true ), {
...
+
+
     run( wiki, linkElement ) {
-        console.log( linkElement );
+	if ( linkElement !== null ) {
+	// Find result container
+        const element = crawlUntilParentFound( linkElement, this.ENGINE_RESULT_SELECTOR );
+
+	const isTopLevel = (a) => {
+	    return a.href.startsWith(
```
Comment (alex4401) [inline js/ddg.js:176]: Don't need the line break. This can fit within 120 columns.

File: js/ddg.js [hunk 3]
```diff
@@ -14,19 +15,223 @@ const wikis = prepareWikisInfo( getWikis( false, true ), {
...
+	};
+	
+	// Check all a relements in the result and verify that it's a link to the same wiki
+	
+	for ( const a of element.getElementsByTagName('a')) {
+	    if(isTopLevel(a))
+		break;
+	}
+	
+        if ( !this.isLocked(element) ) {
```
Comment (alex4401) [inline js/ddg.js:187]: Spacing.

File: js/ddg.js [hunk 4]
```diff
@@ -14,19 +15,223 @@ const wikis = prepareWikisInfo( getWikis( false, true ), {
...
+
+	    
+	    // Add a redirected badge
+            if ( element !== null ) {
+                const titleHeader = element.querySelector( `.${this.SPAN_TITLE_ELEMENT_SELECTOR}` );
```
Comment (alex4401) [inline js/ddg.js:192]: This should be already in the constant and not templated here.

File: js/ddg.js [hunk 5]
```diff
@@ -14,19 +15,223 @@ const wikis = prepareWikisInfo( getWikis( false, true ), {
...
+                if ( titleHeader ) {
+                    titleHeader.parentElement.appendChild( this.makeBadgeElement( isTopLevel ) );
+                    this.lock( titleHeader );
+                }
+		// Rewrite href links
+		for ( const a of element.getElementsByTagName("a")) {
```
Comment (alex4401) [inline js/ddg.js:198]: Quotes and spacing.

File: js/ddg.js [hunk 6]
```diff
@@ -14,19 +15,223 @@ const wikis = prepareWikisInfo( getWikis( false, true ), {
...
+	    }
+        }
+	}
     }
 };
 
 
-invokeSearchModule( wikis, rewrite.run.bind( rewrite ), filter.run.bind( filter ) );
+function observedRun() {
+    if ( document.querySelector('#react-layout'))
```
Comment (alex4401) [inline js/ddg.js:230]: Braces and spacing.

File: js/ddg.js [hunk 7]
```diff
@@ -14,19 +15,223 @@ const wikis = prepareWikisInfo( getWikis( false, true ), {
...
+	if (document.querySelector(rewrite.URL_ELEMENT_SELECTOR) > 0) {
+	    
+	}
+	invokeSearchModule( wikis, rewrite.run.bind( rewrite ), filter.run.bind( filter ) );
+}
+
+assembleMutationObserver(observedRun);
```
Comment (alex4401) [inline js/ddg.js:237]: Spacing.

File: js/ddg.js [hunk 8]
```diff
@@ -14,19 +15,223 @@ const wikis = prepareWikisInfo( getWikis( false, true ), {
...
 };
 
 
-invokeSearchModule( wikis, rewrite.run.bind( rewrite ), filter.run.bind( filter ) );
+function observedRun() {
+    if ( document.querySelector('#react-layout'))
+	if (document.querySelector(rewrite.URL_ELEMENT_SELECTOR) > 0) {
+	    
+	}
+	invokeSearchModule( wikis, rewrite.run.bind( rewrite ), filter.run.bind( filter ) );
```
Comment (alex4401) [inline js/ddg.js:234]: I think you intended to put this line inside the if directly above it?

File: js/ddg.js [hunk 9]
```diff
@@ -14,19 +15,223 @@ const wikis = prepareWikisInfo( getWikis( false, true ), {
...
+                    titleHeader.parentElement.appendChild( this.makeBadgeElement( isTopLevel ) );
+                    this.lock( titleHeader );
+                }
+		// Rewrite href links
+		for ( const a of element.getElementsByTagName("a")) {
+		    this.rewriteLink( wiki, a );
+		}
+                // Rewrite title
+                for ( const span of element.getElementsByTagName( 'span' ) ) {
+		     if( !span.classList.contains( this.SPAN_TITLE_ELEMENT_SELECTOR ) ) {
```
Comment (alex4401) [inline js/ddg.js:203]: Indenting.

**Review by alex4401 — CHANGES_REQUESTED** [review]
Initial review done, haven't tested, went a little harder on you as I know you. Fine to respond over Discord but please mark individual comments as addressed.

File: js/ddg.js [hunk 10]
```diff
@@ -14,19 +15,216 @@ const wikis = prepareWikisInfo( getWikis( false, true ), {
...
+	    const element = linkElement.closest( 'article' );
+ 
+	    const isTopLevel = (a) => {
+		return a.href.startsWith( `https://${wiki.oldId || wiki.id}.fandom.com` );
+	    };
+	
+            if ( element !== null ) {
+		// Check all a relements in the result and verify that it's a link to the same wiki
+		for ( const a of element.getElementsByTagName( 'a' ) ) {
+		    if( isTopLevel( a ) )
```
Comment (alex4401) [inline js/ddg.js:186]: This does nothing?
